torch/utils/checkpoint.py

```diff
@@ -126,30 +126,33 @@ def get_device_type() -> str:
 
 
 def _infer_device_type(*args):
-    device_types = []
+    devices = []
 
-    def add_device_types(arg):
-        nonlocal device_types
+    def add_device(arg):
+        nonlocal devices
         if isinstance(arg, torch.Tensor) and not arg.device.type == "cpu":
-            device_types.append(arg.device.type)
-    tree_map(add_device_types, args)
+            devices.append(arg.device)
+    tree_map(add_device, args)
 
-    device_types_set = set(device_types)
+    return _infer_device_type_from_devices(devices)
+
+def _infer_device_type_from_devices(devices):
```

Comment:
I initially thought to rename above one as `_infer_device_type_from_tensors`, but it occurs it's used in some tests. So, to minimize the change I left it intact.